### PR TITLE
Ensure unique heading links

### LINKED ISSUE #368: URL hash collide when different subsections have the same title
### Is this a bug report?

Yes

### Have you read the [Contributing Guidelines](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md)?

Yes

### Environment

Doesn't matter

### Steps to Reproduce

Try any markdown file

```md
# foo
## baz

# bar
## baz
```

Maybe `baz-1`. for the second one is good enough? No way to guarantee the titles from not clashing, since you might actually have a title called `baz 1`
  
### Expected Behavior

Not sure

### Actual Behavior

In this case, both `baz` will render to html with anchor `baz`, so `my-page#baz` points to the first one always.

### PR DESCRIPTION
**Pull request by SleepWalker**
## Motivation

* ensure, that each heading has it's unique anchor per document (fixes #368)
* use the same logic to generate hash links for `main content`, `OnPageNav` and `AUTOGENERATED_TABLE_OF_CONTENTS`
* Fix filtering in  `OnPageNav` so that by default it collects headings of second and third level
* Build `AUTOGENERATED_TABLE_OF_CONTENTS` using `mardown-toc` parsing results

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I've wrote some tests, that should be enough on my opinion.

To check the fixes manually you need write the doc file with the following lines:

```
## foo
### foo
### foo 1
## foo 1
## foo 2
### foo
```

By rendering on master branch you will get following anchor links:

```
#foo
#foo
#foo-1
#foo-1
#foo-2
#foo
```

After this PR applied:

```
#foo
#foo-1
#foo-1-1
#foo-1-2
#foo-2
#foo-3
```

The anchor indexes must not change/increment after page refresh.

Repeat the same with `<AUTOGENERATED_TABLE_OF_CONTENTS>` and third level headings

### Clashing with headings ending with number

chenglou wrote:

> Maybe baz-1. for the second one is good enough? No way to guarantee the titles from not clashing, since you might actually have a title called baz 1

I have checked how github renders in such cases. It adds another one suffix, e.g.: `baz-1-1`.

Alternatively we can add `--1` instead of `-1`. In this case we will never have any clashing, because toSlug() function will never generate a slug with two hyphens.

Should we mimic github's behavior or take an easier approach with double hyphen? (initially in this PR I have implemented github's behavior)

### `<Header>`, that uses `toSlug`

I've found `<Header>`, but I can't find where it is being used. Should I leave it as it is?

### What the minimal node version do we support?

Probably I not that attentive, but I've found no info regarding the node version and no babel too. So it is a little bit scary to use all the new language features. In some places we are using template string, in other - not and so on.

### CODE REVIEW

**Comment by SleepWalker** [comment]
Note, that this PR depends from #492

**Comment by microbouji** [comment]
This is great! I'll look into the code some more later, but to quickly answer a few things:

> All the users already have TOCs, that pickup h4 headings by default (and h5, and h6 and so on). Should I break the things for them, but fix a bug? Or should we by default allow all the headings in?

Don't know how it should work in `AUTOGENERATED_TABLE_OF_CONTENTS`. In the on-page nav that's a regression from latest release that should be fixed ( that no-op filter :) ).

The idea is the nav produces at most two unordered lists that by default only include `h2`s in the top-level `ul` and `h3`s in the nested `ul`. We give users the option to change which headings show up in the top and sub-level `ul`s, through an `onPageNavHeadings` option as described here: https://github.com/facebook/Docusaurus/pull/475. It is not documented yet; we expect everyone who's using the on-page nav to currently rely on the default of `h2`s and `h3`s in there (except for that regression as mentioned above), so if you have any ideas here it should be possible to still change it.

### A more general `getTOC`

I wrote the getTOC function when implementing the on-page nav. It's not as general as its name might suggest. If it's going to be used for `AUTOGENERATED_TABLE_OF_CONTENTS` as well, it's probably best to simlpify it like this:

```
in: doc mardown
out: flat array with *ALL* heading objects each with title, unique hash link and level
```

So basically mardown-toc's json result with our toSlug and anchors applied. Then leave the extra logic to the part that needs it:

* `AUTOGENERATED_TABLE_OF_CONTENTS` can just loop through them all (or up to h4, however that works) and print. In fact, if we don't care about that edge case I found with the unique slugs, we can even directly register markdown-toc as a remarkable plugin for this.
* `<OnPageNav>` can loop through them and build that two-levels structure with children properties.

Sorry this turned into an essay :)

**Comment by SleepWalker** [comment]
With `getTOC` there is no problems. It currently satisfies all the requirements.

Regarding regression with heading level filtering I've fixed it :)

Thankfully to getTOC `AUTOGENERATED_TABLE_OF_CONTENTS` is more stable too.

I'll update the description with more detailed changlelog

File: lib/server/generate.js [hunk 1]
```diff
@@ -41,15 +41,12 @@ function execute() {
   // takes the content of a doc article and returns the content with a table of
   // contents inserted
   const insertTableOfContents = rawContent => {
-    const regexp = /\n###\s+(`.*`.*)\n/g;
-    let match;
-    const headers = [];
-    while ((match = regexp.exec(rawContent))) {
-      headers.push(match[1]);
-    }
+    const filterRe = /`[^`]*`/;
```
Comment (microbouji) [inline lib/server/generate.js]: Anchor this with both `^` and `$` so it only matches when the backticks wrap the full heading like the old regexp.

Comment (SleepWalker) [reply lib/server/generate.js]: Ok. I'll do that, but the old one can match the following:

```
###     `foo` bar baz
```

This is why I've decided, that the match should not be strict

Comment (microbouji) [reply lib/server/generate.js]: Only anchor the beginning then. Trying to infer the intention of the old one.

**Comment by SleepWalker** [comment]
@JoelMarcey the updates are pushed

**Comment by JoelMarcey** [comment]
@SleepWalker this looks great. I am about ready to merge it. Just a couple of comments...

> Should we mimic github's behavior or take an easier approach with double hyphen? (initially in this PR I have implemented github's behavior)

I am fine with GitHub's behavior on this.

cc @chenglou 

> What the minimal node version do we support?

Docusaurus requires at least node 6. https://docusaurus.io/docs/en/installation.html

Is that reasonable?

> I've found <Header>, but I can't find where it is being used. Should I leave it as it is?

Can you explain this a little more? Not sure what this is referring to in order to answer your question.

**Comment by SleepWalker** [comment]
> Docusaurus requires at least node 6. https://docusaurus.io/docs/en/installation.html
>
> Is that reasonable?

It's ok. This was enough for my question :)

> Can you explain this a little more? Not sure what this is referring to in order to answer your question.

I'm talking about this component:

https://github.com/facebook/Docusaurus/blob/master/lib/core/Header.js

This component uses `toSlug`, but I can't find where this component is required, to check, that `toSlug` is enough. Should this component be removed?

File: lib/core/__tests__/anchors.tests.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+test('Anchors rendering', () => {
+  expect(
+    render([{hLevel: 1}, {content: 'Hello world'}], 0, {}, {})
+  ).toMatchSnapshot();
+  expect(
+    render([{hLevel: 2}, {content: 'Hello small world'}], 0, {}, {})
+  ).toMatchSnapshot();
+});
+
+test('Each anchor is unique across rendered document', () => {
```
Comment (yangshun) [inline lib/core/__tests__/anchors.tests.js]: Could you also add in a test for the following and ensure it works well?

```
# foo
# foo 1
# foo 
# foo 1
```

File: lib/core/toSlug.js [hunk 3]
```diff
@@ -18,7 +18,16 @@ const exceptAlphanum = new RegExp(
   'g'
 );
 
-module.exports = string => {
+/**
+ * Converts a string to a slug, that can be used in heading anchors
+ *
+ * @param  {string} string
+ * @param  {Object} [context={}] - an optional context to track used slugs and
+ *                                 ensure, that new slug will be unique
```
Comment (yangshun) [inline lib/core/toSlug.js]: There shouldn't be a comma after ensure

**Comment by yangshun** [comment]
@SleepWalker @microbouji It seems that `markdown-toc` has built-in support for repeated headings, as [seen here](https://github.com/jonschlinkert/markdown-toc#highlights) and [here](https://gist.github.com/jonschlinkert/ac5d8122bfaaa394f896). Is it not working for us? Would like to avoid reinventing the wheel if possible.

**Comment by microbouji** [comment]
@yangshun yeah that's the issue we had with markdown-toc's repeated headings - it doesn't support headings ending in numbers like this: https://github.com/facebook/Docusaurus/pull/492#issuecomment-381551904.

It's a tradeoff I guess between supporting this use case, and making the code less complex by just relying on markdown-toc.

**Comment by SleepWalker** [comment]
@yangshun done
